records/2024-06-04.manifests-kustomize-cicd.md

```diff
@@ -0,0 +1,140 @@
+
+This issue is a problem regardless of whether we use Kustomize or Helmfile.
+
+#### Helmfile Image Versions
+
+For helmfile releases, we are relying on the helmfile logic to set the image versions in the overrides. For example, we may have something like the code block below, which hard codes the tag for production, but accepts either an environment variable or will default to "latest" if it does not exist.
+
+```yaml
+      {{ if eq .Environment.Name "production" }}
+      image:
+        tag: release1.24
+      {{ else }}
+      image:
+        tag: {{ env "DOCKER_TAG" | default "latest" }}
+      {{ end }}
+```
+
+This is an ok solution, but it's not ideal to have to modify the individual override files to set image versions for production manually. It would be ideal to have this done automatically like with the PR bot.
+
+## Options
+
+Because Helmfile has been identified as a long term goal to migrate GC Notify to, the options below will all involve moving the Kustomize code to Helmfile.
+
+### Option 1: Migrate Everything to Helmfile Keeping the Existing CI/CD Processes
+
+The first option is to simply move Notify to helmfile, and adapt the existing CI/CD process. This will be a relatively easy change as most of this work has been completed in a manifests branch. The Notify PR bot will have to be modified to change the image tag in the overrides files of helmfile instead of the Kustomization.yaml file.
+
+The staging deployments would remain exactly the same as they are now.
+
+#### Diagram
+
+![Production CICD Option 1](./diagrams/2024-06-04.manifests-current-cicd/cicd-option1.png)
+
+#### Pros
+
+- Least effort
+- Consolidates all Kubernetes resources to a single technology
+- Introduces the ability to add "Diff" previews to Notify Applications on manifests PRs
+
+#### Cons
+
+- Does not solve the release process differences between staging and production
+
+### Option 2: Per-Environmnent Centralized Image Version File
+
+The second option is to try and overhaul the CI/CD process in staging to make it more consistent with production, while also maintaining the flexibility of rapid deployments to staging. We could create a github action in each notification-x repo that automatically modifies a centralized <environment>_image_versions.json file. When this change is merged to main, a manifests job would trigger a new deploy to either staging or production.
+
+#### Diagram
+
+![Staging/Production CICD Option 2](./diagrams/2024-06-04.manifests-current-cicd/cicd-option2a.png)
+
+#### Pros
+
+- Consolidates all Kubernetes resources to a single technology
+- Introduces the ability to add "Diff" previews to Notify Applications on manifests PRs
+- Staging and Production have the same release process
+
+#### Cons
+
+- More complex - will need to enable/disable branch protection when updating the image.json
```

Comment:
This suggest leaving the PR bot behind, right? I think we want to keep it somehow, but potentially updating the `image.json` file instead. Because it adds context to the changes, such as titles and authors. It could generate a changelog for each releases as well (which it does not at the moment but the point being that this adds potential to documenting and auditing our release process).